### ISSUE TITLE
YOLOv3

### ISSUE THREAD

**Issue by JJAndy** [liangheming/gfocal#1]
how to employ gfocal loss  to yolov3

**Comment by liangheming** [comment]
it's easy to employ QFL to classification branch in YOLOv3,simply use 'classification-IoU joint representation(describled in Generalized Focal Loss )' as YOLO classification targets,then apply function QFL
```
class QFL(object):
    def __init__(self, beta=2.0):
        super(QFL, self).__init__()
        self.beta = beta
        self.bce = torch.nn.BCEWithLogitsLoss(reduction="none")

    def __call__(self, predicts, targets):
        pt = predicts.sigmoid()
        loss = self.bce(predicts, targets) * ((targets - pt).abs().pow(self.beta))
        return loss
```

but apply DFL to box branch is not that obvious, One way is to change the box branch (cxcy_wh in YOLOv3) to FCOS style (ltrb).then apply DFL

**Comment by JJAndy** [comment]
ok, thank you so much. so, it mean that I need to change the head of yolov3?



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "liangheming"<notifications@github.com&gt;; 
发送时间: 2020年12月8日(星期二) 下午3:02
收件人: "liangheming/gfocal"<gfocal@noreply.github.com&gt;; 
抄送: "你头发乱了喔"<3214332982@qq.com&gt;; "Author"<author@noreply.github.com&gt;; 
主题: Re: [liangheming/gfocal] YOLOv3 (#1)





 
it's easy to employ QFL to classification branch in YOLOv3,simply use 'classification-IoU joint representation(describled in Generalized Focal Loss )' as YOLO classification targets,then apply function QFL
 class QFL(object):     def __init__(self, beta=2.0):         super(QFL, self).__init__()         self.beta = beta         self.bce = torch.nn.BCEWithLogitsLoss(reduction="none")     def __call__(self, predicts, targets):         pt = predicts.sigmoid()         loss = self.bce(predicts, targets) * ((targets - pt).abs().pow(self.beta))         return loss  
but apply DFL to box branch is not that obvious, One way is to change the box branch (cxcy_wh in YOLOv3) to FCOS style (ltrb).then apply DFL

**Comment by liangheming** [comment]
Anyway, it's a choice.but also you can directly model the box regression target (delta 'cxcy_wh'  in YOLOv3)  as a distribution. if you do so,there are some details you should know.
**value range** , box regression target(delta 'ltrb') used in gfocal is normalized by grid scale which makes the regression targets range from 0 to *reg_max* +1.
